test/unit/destructuring.spec.ts

```diff
@@ -140,3 +140,31 @@ describe("array destructuring optimization", () => {
             .expectToMatchJsResult();
     });
 });
+
+// TODO: Adjust this test to use expectToMatchJsResult() and testCases when TypeScript can correctly handle the code.
+// See Babel's transpiler for how this code is meant to be transpiled.
+test.each([
+    ["foo", "['bar']"],
+    ["[foo]", "[['bar']]"],
+    ["[foo = 'bar']", "[[]]"],
+    ["{ foo }", "[{ foo: 'bar' }]"],
+    ["{ x: foo }", "[{ x: 'bar' }]"],
+    ["{ foo = 'bar' }", "[{}]"],
+])("forof assignment updates dependencies", (initializer, expression) => {
+    util.testModule`
+        let foo = '';
+        export { foo };
+        for (${initializer} of ${expression} as any) {}
```

Comment:
Is it possible to avoid casting to any? That wouldn't work if we'll start using type information in destructuring, i.e. to detect iterables